### PR TITLE
feat(core): inspect persisted conversation branches

### PR DESCRIPTION
**Pull request by doudouOUC**
## What this PR does

Adds a read-only conversation topology inspector that identifies semantic leaves in persisted session trees, reports missing-parent, cycle, and conflicting-parent diagnostics, and produces deterministic summaries for each branch.

Normalizes metadata-only terminal records so custom-title and session-artifact writes do not appear as phantom conversation branches, while preserving recovery-significant system records such as rewind, compression, attribution, and file-history checkpoints.

Classifies rewind descendants and rewind siblings strictly as structural relationships. It does not select an active branch, mark a sibling abandoned, mutate the transcript, or change current resume, fork, daemon, ACP, or CLI behavior.

## Why it's needed

Persisted sessions already form a tree through parent UUID links, but restart currently reconstructs one physically selected tail. Rewinds or concurrent writers can leave multiple valid sibling histories in the same JSONL file, so selecting only the physical tail can hide recoverable work. This PR provides the independent read-side foundation needed to enumerate and summarize those branches before a later write-side change adds exclusive ownership and durable branch selection.

### How to verify

1. Run `cd packages/core && npx vitest run src/utils/conversation-branches.test.ts src/utils/conversation-chain.test.ts`; expect all 28 tests to pass.
2. Confirm metadata-only title and artifact tails collapse without removing significant system checkpoints.
3. Confirm both sides of a rewind fork remain visible: the rewind path is classified as a rewind descendant, while the sibling is classified structurally and is not discarded.
4. Confirm the sanitized incident regression reduces three raw terminal records to two semantic conversation branches and that every returned leaf is accepted by the existing explicit reconstruction path.

### Evidence (Before & After)

N/A — this PR adds a pure read-side core API and does not change user-visible behavior.

### Tested on

|     OS     | Status |
| :--------: | :----: |
|  🍏 macOS  |   ✅   |
| 🪟 Windows |  N/A   |
|  🐧 Linux  |  N/A   |

### Environment (optional)

Node.js v22.22.3, local non-sandboxed unit tests.

## Risk & Scope

- Main risk or tradeoff: semantic leaf normalization intentionally uses a narrow allowlist; unknown and recovery-significant system records remain visible rather than being guessed away.
- Not validated / out of scope: active-branch selection, writer ownership, persistence, and UI integration are intentionally deferred. Full-workspace typecheck is currently blocked by pre-existing Ink type export errors under the CLI selection code; core typecheck, lint, targeted tests, and the repository build pass.
- Breaking changes / migration notes: none; the new API has no runtime consumer in this PR and existing behavior is unchanged.

## Linked Issues

N/A

<details>
<summary>中文说明</summary>

## 本 PR 做了什么

新增纯读的会话拓扑检查能力，用于识别持久化会话树中的语义叶子、报告父节点缺失、父链循环和重复 UUID 父节点冲突，并为每个分支生成确定性摘要。

对仅包含元数据的终端记录进行归一化，避免自定义标题和会话 artifact 写入被误识别为虚假会话分支，同时保留 rewind、压缩、归因和文件历史检查点等对恢复有意义的 system 记录。

将 rewind descendant 和 rewind sibling 严格定义为结构关系。本 PR 不选择 active branch，不把 sibling 标记为 abandoned，不修改 transcript，也不改变现有 resume、fork、daemon、ACP 或 CLI 行为。

## 为什么需要

持久化会话已经通过父 UUID 链形成树结构，但重启恢复目前只重建一个按物理位置选出的尾部。rewind 或并发 writer 可能在同一 JSONL 文件中留下多个有效 sibling history，只选择物理尾部会隐藏仍可恢复的工作。本 PR 提供独立的读侧基础能力，用于枚举和摘要这些分支；后续写侧变更再补充独占所有权和持久化分支选择。

### 如何验证

1. 运行 `cd packages/core && npx vitest run src/utils/conversation-branches.test.ts src/utils/conversation-chain.test.ts`，预期 28 个测试全部通过。
2. 确认仅包含标题和 artifact 的元数据尾部会折叠，同时不会移除重要的 system 检查点。
3. 确认 rewind 分叉的两侧都保留：rewind 路径被分类为 rewind descendant，sibling 只做结构分类且不会被丢弃。
4. 确认脱敏事故回归夹具会把三个原始 terminal 归一化为两个语义会话分支，并且每个返回叶子都能被现有显式重建链路接受。

### 证据（Before & After）

N/A——本 PR 新增纯读 core API，不改变用户可见行为。

### 测试平台

|     OS     | 状态 |
| :--------: | :--: |
|  🍏 macOS  |  ✅  |
| 🪟 Windows | N/A  |
|  🐧 Linux  | N/A  |

### 环境（可选）

Node.js v22.22.3，本地无 sandbox 单元测试。

## 风险与范围

- 主要风险或取舍：语义叶子归一化刻意使用较窄的白名单；未知或对恢复有意义的 system 记录会保留，不做猜测性删除。
- 未验证或范围外：active branch 选择、writer 所有权、持久化和 UI 接入均有意延后。全 workspace typecheck 当前被 CLI selection 代码中既有的 Ink 类型导出错误阻断；core typecheck、lint、定向测试和仓库 build 均通过。
- 破坏性变更或迁移说明：无；新 API 在本 PR 中没有运行时消费者，现有行为不变。

## 关联 Issue

N/A

</details>

### CODE REVIEW

**Comment by doudouOUC** [comment]
Thanks — no code change from this review. The split is intentional: this PR validates a pure, deterministic read model against the existing persisted ChatRecord schema and a sanitized incident topology, while the later restart/selection change owns mutation and recovery policy. Keeping those concerns separate narrows the risk and lets the topology rules be reviewed without write-side behavior; the later consumer can adapt at the call boundary without changing persisted-data analysis. Maintainer guidance on whether to land this foundation separately is welcome.

File: packages/core/src/utils/conversation-branches.ts [hunk 1]
```diff
@@ -0,0 +1,527 @@
...
+    if (record.type === 'tool_result') {
+      recordCounts.toolResult++;
+    } else {
+      recordCounts[record.type]++;
+    }
+  }
+
+  const firstDistinctIndex = branchPointUuid
+    ? chain.indexOf(branchPointUuid) + 1
+    : 0;
```
[the remark this answers is not in the document]
Comment (doudouOUC) [reply packages/core/src/utils/conversation-branches.ts:364]: Fixed in 69c2912aa9. Added an explicit single-chain test for branchPointUuid: null and the first user summary.

File: packages/core/src/utils/conversation-branches.ts [hunk 2]
```diff
@@ -0,0 +1,527 @@
...
+        leafUuid,
+        chain,
+        rewindUuid,
+        forestPositions,
+        index.firstByUuid,
+      ),
+  );
+
+  const firstRecord = index.firstByUuid.get(chain[0])!;
+  const leafRecord = index.firstByUuid.get(leafUuid)!;
```
[the remark this answers is not in the document]
Comment (doudouOUC) [reply packages/core/src/utils/conversation-branches.ts]: Fixed in 69c2912aa9. Semantic leaves now retain the last collapsed physical leaf for updatedAt, with a regression test and documented contract.

File: packages/core/src/utils/conversation-branches.ts [hunk 3]
```diff
@@ -0,0 +1,527 @@
...
+  const recordCounts = { user: 0, assistant: 0, toolResult: 0, system: 0 };
+  for (const uuid of chain) {
+    const record = index.firstByUuid.get(uuid)!;
+    if (record.type === 'tool_result') {
+      recordCounts.toolResult++;
+    } else {
+      recordCounts[record.type]++;
+    }
```
[the remark this answers is not in the document]
Comment (doudouOUC) [reply packages/core/src/utils/conversation-branches.ts:359]: Fixed in 69c2912aa9. Added a branch-chain regression test that asserts a non-zero toolResult count.

**Comment by doudouOUC** [comment]
Addressed all three review suggestions in `69c2912aa9`.

| Feedback | Action |
| --- | --- |
| `tool_result` count coverage | Added a non-zero `toolResult` regression test. |
| Linear branch with no fork | Added explicit `branchPointUuid: null` and first-user-summary assertions. |
| `updatedAt` after neutral-tail collapse | Retained the last collapsed physical leaf for recency, added a regression test, and documented the contract. |

Verification: full repository build passed; full workspace typecheck passed; 26 targeted conversation branch/chain tests passed.

File: packages/core/src/utils/conversation-branches.ts [hunk 4]
```diff
@@ -0,0 +1,542 @@
...
+}
+
+function findSemanticLeaves(index: ConversationIndex): SemanticLeaf[] {
+  const candidates = new Set<string>();
+  const physicalLeafBySemanticLeaf = new Map<string, string>();
+  for (const uuid of index.firstByUuid.keys()) {
+    if ((index.childrenByUuid.get(uuid)?.length ?? 0) > 0) continue;
+    const leafUuid = collapseNeutralTail(uuid, index.firstByUuid);
+    candidates.add(leafUuid);
+    physicalLeafBySemanticLeaf.set(leafUuid, uuid);
```
[the remark this answers is not in the document]
Comment (doudouOUC) [reply packages/core/src/utils/conversation-branches.ts:254]: Thanks. I am not applying this suggestion because the input array is the authoritative physical JSONL order throughout this inspector: the first physical occurrence defines a UUID parent, and physical indexes determine deterministic branch order. Here, “last physical terminal” means the last appended/input record that collapses to the semantic leaf, not the maximum timestamp. A reverse-ordered array therefore describes a different physical transcript; selecting by timestamp would silently reorder it and would be wrong under clock rollback, cross-writer clock skew, or malformed timestamps. The current overwrite and test intentionally preserve physical append order.

**Comment by wenshao** [comment]
## Local runtime verification — PR #7185 ✅

I built and ran this PR locally against the **real compiled** code (not just vitest) as a merge reference. Net: **the PR does what it claims — behavior is correct for every realistic transcript, and the tests have real teeth.** One latent edge case is worth an optional follow‑up (F1 below); it is non‑blocking because nothing consumes this API yet.

PR 7185 verification summary

### How I verified
- Checked out PR head `69c2912aa` in an isolated git worktree, then esbuild‑bundled `conversation-branches.ts` + `conversation-chain.ts` into real ESM and drove the **compiled** `inspectConversationBranches` / `buildOrderedUuidChain` directly (`inspectConversationBranches` only has a type‑only import, so the bundle is faithful runtime JS).
- Env note: a full `tsc --build` OOMs in my sandbox, but vitest transpiles and runs the real TS source fine — type + behavior are exercised end‑to‑end. Not a PR issue.

### Results
| Check | Result |
|---|---|
| Unit suite (both files) | ✅ **26 passed** (branches 18 + chain 8) |
| Runtime harness on the real compiled fn | ✅ **32 / 32** assertions |
| Fuzz — 5000 faithful append‑only transcripts | ✅ invariant holds (10,390 branches · 2,873 multi‑branch · 0 crashes) |
| Mutation testing | ✅ suite catches regressions (3 / 5 / 1 fails) |

Confirmed at runtime, **beyond** the test file:
- Metadata‑only title/artifact tails collapse; significant system checkpoints (`rewind` / `chat_compression` / `attribution_snapshot` / `file_history_snapshot` / `slash_command`) stay visible.
- Rewind fork keeps **both** sides — `rewind-descendant` vs `rewind-sibling` labelled correctly, neither discarded.
- Sanitized‑incident: 3 raw terminals → 2 semantic branches.
- Diagnostics: `missing-parent` · `parent-cycle` · `conflicting-parent`.
- Summary hygiene: assistant `thought` parts and tool‑call args never leak into summaries; synthetic prompts (`notification` / `cron` / `mid_turn_user_message` / external‑input) are filtered; text is whitespace‑normalized and truncated at 200.
- **Cross‑check invariant** — over 5000 transcripts generated with the *real* writer semantics (active‑tail advance, `updateActiveTail:false` metadata tails, rewind re‑rooting, concurrent‑writer siblings), every returned leaf reconstructs **gap‑free** via `buildOrderedUuidChain` and ends at that leaf, and the branch set is always an antichain.

Mutation testing (proves the suite isn't vacuous):

| Mutation of `conversation-branches.ts` | Tests failed |
|---|---|
| `findBranchPoint` `> 1` → `>= 1` | 3 |
| `collapseNeutralTail` (disable neutral collapse) | 5 |
| `updatedAt` physicalLeaf → semanticLeaf (undo the review‑feedback fix) | 1 |
| baseline restored | 18 passed |

**Minor:** the reviewer test‑plan step 1 says "expect all **23** tests"; head actually has **26** (the review‑feedback commit added 3). Harmless — just bump the number.

### Finding F1 — metadata‑only record with no non‑neutral ancestor → phantom "ordinary" branch (latent, non‑blocking)

`collapseNeutralTail` only collapses a neutral tail into a **non‑neutral ancestor**. When a neutral record has none — e.g. it is the session's first record (`parentUuid = null`) — it is returned **as** the leaf: the exact "phantom conversation branch" the PR says it prevents. For the two artifact subtypes it is worse: `buildOrderedUuidChain` cannot reconstruct that leaf (`selectTranscriptLeaf` rejects artifact records) → **empty chain**, which breaks the "every returned leaf is accepted by the existing reconstruction path" property.

| Root record subtype | `inspectConversationBranches` | `buildOrderedUuidChain` |
|---|---|---|
| `session_artifact_event` | phantom `ordinary` branch | ❌ empty chain — cannot reconstruct |
| `session_artifact_snapshot` | phantom `ordinary` branch | ❌ empty chain — cannot reconstruct |
| `custom_title` | phantom `ordinary` branch | ✅ reconstructs `["meta"]` |

Minimal repro:
```js
const recs = [{ uuid:'meta', parentUuid:null, sessionId:'s', timestamp:'2026-01-01T00:00:00.000Z',
                type:'system', subtype:'session_artifact_event', cwd:'/w', version:'0' }];
inspectConversationBranches(recs).branches;
//   => [{ leafUuid:'meta', classification:'ordinary', ... }]      // phantom branch
buildOrderedUuidChain(recs, { leafUuid:'meta', detectGaps:true }); // => { uuids:[], gaps:[] }  // cannot reconstruct
```

**Reachability:** the only writer that emits artifact records is ACP `session/persist-artifact` (`acpAgent.ts`); a client persisting an artifact before the first prompt would produce a root artifact. I did **not** prove that end‑to‑end, and there is no runtime consumer of `inspectConversationBranches` yet, so this stays latent. Root cause: the inspector indexes **all** records while `buildOrderedUuidChain` excludes artifact subtypes via `isTranscriptConversationRecord`, so the two use different record‑eligibility predicates. Suggested guard for the future write side: drop metadata‑only branches that lack a conversation ancestor, or share one eligibility predicate between the inspector and the reconstruction path so they can never disagree.

### Recommendation
👍 Good to merge as the read‑side foundation. Treat **F1** as a small follow‑up (or a cheap guard) before anything starts selecting/consuming these leaves.

<details>
<summary>🇨🇳 中文说明</summary>

## 本地真实构建验证 — PR #7185 ✅

作为合并参考，我在本地用**真实编译产物**（不仅是 vitest）构建并运行了本 PR。结论：**PR 行为与描述一致 —— 对所有真实 transcript 都正确，且测试有实际约束力。** 有一个潜在边界情况值得作为可选后续处理（下方 F1），但不阻塞合并，因为目前还没有任何运行时消费该 API。

### 验证方式
- 在独立 git worktree 检出 PR head `69c2912aa`，用 esbuild 把 `conversation-branches.ts` + `conversation-chain.ts` 打包为真实 ESM，直接驱动**编译后的** `inspectConversationBranches` / `buildOrderedUuidChain`（`inspectConversationBranches` 只有 type‑only import，打包产物即忠实的运行时 JS）。
- 环境说明：完整 `tsc --build` 在我的 sandbox 里 OOM，但 vitest 能正常转译并运行真实 TS 源码，类型与行为已端到端覆盖 —— 与 PR 无关。

### 结果
| 检查项 | 结果 |
|---|---|
| 单测（两个文件） | ✅ **26 通过**（branches 18 + chain 8） |
| 真实编译产物 runtime 夹具 | ✅ **32 / 32** 断言 |
| 模糊测试 —— 5000 个忠实 append‑only transcript | ✅ 不变量成立（10,390 分支 · 2,873 多分支 · 0 崩溃） |
| 变异测试 | ✅ 测试能抓到回归（3 / 5 / 1 失败） |

在测试文件**之外**的运行时确认：
- 仅含元数据的 title/artifact 尾部会折叠；重要 system 检查点（`rewind` / `chat_compression` / `attribution_snapshot` / `file_history_snapshot` / `slash_command`）保留可见。
- rewind 分叉两侧都保留 —— `rewind-descendant` 与 `rewind-sibling` 分类正确，两侧都不被丢弃。
- 脱敏事故场景：3 个原始 terminal → 2 个语义分支。
- 诊断：`missing-parent` · `parent-cycle` · `conflicting-parent`。
- 摘要卫生：assistant 的 `thought` 片段与 tool‑call 参数不会泄漏进摘要；合成 prompt（`notification` / `cron` / `mid_turn_user_message` / 外部输入）被过滤；文本做空白归一并在 200 处截断。
- **交叉校验不变量** —— 在 5000 个按*真实* writer 语义生成的 transcript（active‑tail 前进、`updateActiveTail:false` 元数据尾、rewind 重新定根、并发 writer sibling）上，每个返回的叶子都能通过 `buildOrderedUuidChain` **无 gap** 重建并终止于该叶子，且分支集合始终是 antichain。

变异测试（证明测试非空洞）：

| 对 `conversation-branches.ts` 的变异 | 失败测试数 |
|---|---|
| `findBranchPoint` `> 1` → `>= 1` | 3 |
| `collapseNeutralTail`（关闭中性折叠） | 5 |
| `updatedAt` physicalLeaf → semanticLeaf（回退 review 修复） | 1 |
| 恢复 baseline | 18 通过 |

**小问题：** reviewer 测试计划第 1 步写的是「23 个测试」，但 head 实际是 **26 个**（review‑feedback commit 新增了 3 个）。无害，更新下数字即可。

### 发现 F1 —— 无非中性祖先的纯元数据记录 → 幽灵 "ordinary" 分支（潜在，不阻塞）

`collapseNeutralTail` 只会把中性尾部折叠到**非中性祖先**。当中性记录没有这样的祖先时 —— 例如它是会话的第一条记录（`parentUuid = null`）—— 它会被当作叶子返回，正是 PR 声称要避免的「幽灵会话分支」。对两个 artifact 子类型更糟：`buildOrderedUuidChain` 无法重建该叶子（`selectTranscriptLeaf` 拒绝 artifact 记录）→ **空链**，破坏了「每个返回叶子都能被现有重建链路接受」这一性质。

| 根记录子类型 | `inspectConversationBranches` | `buildOrderedUuidChain` |
|---|---|---|
| `session_artifact_event` | 幽灵 `ordinary` 分支 | ❌ 空链 —— 无法重建 |
| `session_artifact_snapshot` | 幽灵 `ordinary` 分支 | ❌ 空链 —— 无法重建 |
| `custom_title` | 幽灵 `ordinary` 分支 | ✅ 重建为 `["meta"]` |

最小复现：
```js
const recs = [{ uuid:'meta', parentUuid:null, sessionId:'s', timestamp:'2026-01-01T00:00:00.000Z',
                type:'system', subtype:'session_artifact_event', cwd:'/w', version:'0' }];
inspectConversationBranches(recs).branches;
//   => [{ leafUuid:'meta', classification:'ordinary', ... }]      // 幽灵分支
buildOrderedUuidChain(recs, { leafUuid:'meta', detectGaps:true }); // => { uuids:[], gaps:[] }  // 无法重建
```

**可达性：** 唯一会写 artifact 记录的 writer 是 ACP `session/persist-artifact`（`acpAgent.ts`）；若客户端在首个 prompt 之前持久化 artifact，就会产生根 artifact。我**没有**端到端证明这一点，且目前没有任何运行时消费 `inspectConversationBranches`，所以仅为潜在问题。根因：inspector 索引**全部**记录，而 `buildOrderedUuidChain` 通过 `isTranscriptConversationRecord` 排除 artifact 子类型，两者使用了不同的记录准入判定。后续写侧的建议护栏：丢弃缺少会话祖先的纯元数据分支，或让 inspector 与重建链路共用同一准入判定，避免两者产生分歧。

### 建议
👍 作为读侧基础可以合并。把 **F1** 当作小的后续项（或加个便宜的护栏）即可 —— 在任何逻辑开始选择/消费这些叶子之前处理。

</details>

---
<sub>Verified locally on Linux, Node v22.22.2, against PR head `69c2912aa`. Harness drives the real compiled functions; 5000‑transcript fuzz uses the real chatRecordingService active‑tail/rewind semantics.</sub>

File: packages/core/src/utils/conversation-branches.ts [hunk 5]
```diff
@@ -0,0 +1,542 @@
...
+  const visited = new Set<string>();
+  let currentUuid = leafUuid;
+
+  while (!visited.has(currentUuid)) {
+    visited.add(currentUuid);
+    const record = firstByUuid.get(currentUuid);
+    if (!record || !isNeutralTailRecord(record)) break;
+    const parentUuid = record.parentUuid;
+    if (parentUuid === null || !firstByUuid.has(parentUuid)) break;
+    currentUuid = parentUuid;
```
Comment (wenshao) [inline packages/core/src/utils/conversation-branches.ts:318]: **[Suggestion]** When a neutral-tail record (`session_artifact_event` / `session_artifact_snapshot`) is a physical leaf whose `parentUuid` is `null` or missing, `collapseNeutralTail` breaks on this guard and returns the artifact's own UUID as the semantic leaf. `buildOrderedUuidChain` then rejects that subtype (`selectTranscriptLeaf` → `isTranscriptConversationRecord` → `isTranscriptArtifactRecord`), returning `{ uuids: [], gaps: [] }` for a leaf the inspector reports as reconstructable. — Concrete cost: a corrupted or artifact-only transcript (the artifact record's parent lost) makes the inspector promise a leaf that the reconstruction path silently turns into nothing; the "returns leaves accepted by the existing explicit reconstruction path" test only covers user/assistant records, so this edge is untested. Consider guarding in `findSemanticLeaves` (skip, or emit a diagnostic for, a collapsed UUID that is still a neutral-tail record) before the deferred write-side integration lands.

_— qwen3.8-max-preview via Qwen Code /review_

File: packages/core/src/utils/conversation-branches.ts [hunk 6]
```diff
@@ -0,0 +1,542 @@
...
+  const leaves = [...candidates].filter(
+    (candidate) => !superseded.has(candidate),
+  );
+
+  return leaves
+    .sort(
+      (left, right) =>
+        (index.physicalIndexByUuid.get(left) ?? 0) -
+        (index.physicalIndexByUuid.get(right) ?? 0),
+    )
```
Comment (wenshao) [inline packages/core/src/utils/conversation-branches.ts]: **[Suggestion]** This sort keys on the semantic leaf's own physical index, but when neutral-tail collapsing maps a later physical leaf to an earlier ancestor, that index differs from the collapsed physical leaf's — so branch order diverges from the "physical leaf order" the test suite names. — Concrete cost: with `[C(root, idx 0), A(root, idx 1), B(custom_title→A, idx 2), D(custom_title→C, idx 3)]` the sort yields `[C, A]` (semantic indices 0 < 1) while physical-leaf order (B idx 2 < D idx 3) gives `[A, C]`; a consumer displaying branches in the returned order shows root-creation order rather than latest-append order. The "uses physical leaf order even when timestamps are out of order" test only exercises the case where semantic leaf ≡ physical leaf, so this is untested. Sort by the collapsed physical leaf's index:

```suggestion
    .sort(
      (left, right) =>
        (index.physicalIndexByUuid.get(physicalLeafBySemanticLeaf.get(left)!) ?? 0) -
        (index.physicalIndexByUuid.get(physicalLeafBySemanticLeaf.get(right)!) ?? 0),
    )
```

_— qwen3.8-max-preview via Qwen Code /review_

**Review by wenshao — COMMENTED** [review]
Reviewed — no blockers. Suggestions are inline.

_— qwen3.8-max-preview via Qwen Code /review_

**Comment by doudouOUC** [comment]
@wenshao Thanks for the thorough runtime verification. I independently reproduced F1 on the current head and confirmed that an ACP artifact can be persisted before the first conversation record.

| Item | Action |
| --- | --- |
| F1: metadata-only root exposed as an ordinary branch | Fixed in `0060e0de38`. Neutral terminal chains with no known non-neutral conversation ancestor are now omitted. |
| Regression coverage | Added coverage for root `custom_title`, `session_artifact_event`, and `session_artifact_snapshot` records. |
| Reviewer test-plan count | Updated from 23 to 27 in both English and Chinese. |

Verification: full repository build passed, workspace typecheck passed, and the targeted branch/chain suites passed 27/27.

File: packages/core/src/utils/conversation-branches.ts (hunk 5 above)
Comment (doudouOUC) [reply packages/core/src/utils/conversation-branches.ts:318]: Fixed in `0060e0de38`. Neutral-only terminal chains without a conversation ancestor are now omitted, with regression coverage for all three neutral subtypes.

File: packages/core/src/utils/conversation-branches.ts (hunk 6 above)
Comment (doudouOUC) [reply packages/core/src/utils/conversation-branches.ts]: Fixed in `119a2f673b`. Branch ordering now uses each collapsed physical terminal's index, with a regression covering ancestor order that differs from terminal order.

**Comment by doudouOUC** [comment]
Review follow-up summary for `119a2f673b`:

| Feedback | Action |
| --- | --- |
| Neutral-only artifact/title roots could be reported as unreconstructable branches | Fixed in `0060e0de38`; such roots are omitted and covered by regression tests. |
| Collapsed branches were sorted by semantic ancestor position | Fixed in `119a2f673b`; ordering now follows the actual collapsed physical terminal position. |

Verification: full repository build passed, workspace typecheck passed, and branch/chain unit suites passed 28/28. No feedback was rejected or deferred.

File: packages/core/src/utils/conversation-branches.test.ts [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+
+  it('filters synthetic prompts and thoughts and truncates summary text', () => {
+    const longText = 'x'.repeat(220);
+    const records = [
+      record('real-user', null, {
+        message: { role: 'user', parts: [{ text: '  real\n request  ' }] },
+      }),
+      record('notification', 'real-user', {
+        subtype: 'notification',
+        message: { role: 'user', parts: [{ text: 'synthetic prompt' }] },
```
[the remark this answers is not in the document]
Comment (doudouOUC) [reply packages/core/src/utils/conversation-branches.test.ts:368]: Thanks — agreed and fixed in `07290d8ce2`. The existing summary-hygiene regression now covers `notification`, `cron`, and `mid_turn_user_message` in one linear branch.

**Comment by doudouOUC** [comment]
Review follow-up summary for `07290d8ce2`:

| Feedback | Action |
| --- | --- |
| Synthetic user filtering lacked explicit coverage for `cron` and `mid_turn_user_message` | Fixed by extending the existing summary-hygiene regression to cover both subtypes in the same linear branch. |

Verification: full repository build passed, workspace typecheck passed, and branch/chain unit suites passed 28/28. No feedback was rejected or deferred.

**Comment by wenshao** [comment]
@qwen-code /triage

**Review by wenshao — APPROVED** [review]
